### PR TITLE
gawk: 5.1.1 -> 5.2.1

### PR DESCRIPTION
**Pull request by trofi**
changelogs:
- 5.2.1: https://lists.gnu.org/archive/html/info-gnu/2022-11/msg00008.html
- 5.2.0: https://lists.gnu.org/archive/html/info-gnu/2022-09/msg00003.html

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
@ofborg eval

**Comment by trofi** [comment]
Built `system` successfully. Should be safe for `staging`.

**Comment by trofi** [comment]
Might fail on clang or macos:

```
checking for clang options needed to detect all undeclared functions... cannot detect
configure: error: in `/private/tmp/nix-build-gawk-5.2.1.drv-0/gawk-5.2.1':
configure: error: cannot make clang report undeclared builtins
```

Will have a look.

**Comment by trofi** [comment]
`x86_64-darwin` was fine though. Might be something aarch64-specific.

**Comment by devusb** [comment]
The failure seems to be related to this change in gawk: https://git.savannah.gnu.org/cgit/gawk.git/tree/README_d/README.macosx?h=gawk-5.2.1#n1

Conditionally adding `--disable-pma` made the build complete for me on aarch64-darwin.

**Comment by trofi** [comment]
Great observation! I'll look a bit more into the option if it makes sense to disable it for just this targets for universally and amend the PR.

**Comment by trofi** [comment]
Pulled yet unreleased upstreamed https://git.savannah.gnu.org/cgit/gawk.git/patch/?id=e0b7737930f8a677d3c509f8ce72b9130965ec0a patch. It disables pma for non-x86_64 darwins. Should help as well.

File: pkgs/tools/text/gawk/default.nix [hunk 1]
```diff
@@ -17,18 +18,26 @@ assert (doCheck && stdenv.isLinux) -> glibcLocales != null;
...
+  ];
+
   # When we do build separate interactive version, it makes sense to always include man.
   outputs = [ "out" "info" ]
     ++ lib.optional (!interactive) "man";
 
-  nativeBuildInputs = lib.optional (doCheck && stdenv.isLinux) glibcLocales;
+  nativeBuildInputs = lib.optional (doCheck && stdenv.isLinux) glibcLocales
+    # no-pma fix
+    ++ [ autoreconfHook ];
```
Comment (SuperSandro2000) [inline pkgs/tools/text/gawk/default.nix:40]: normally we would switch those

Comment (trofi) [reply pkgs/tools/text/gawk/default.nix:40]: Good point! When I was writing this I mentally tried to make it easy to remove on next update.

**Comment by alyssais** [comment]
This has also broken the x86_64-unknown-linux-musl stdenv:

```
/nix/store/s22ypd6p0wm9ywhd94fxirvq4bs93d7a-binutils-2.39/bin/ld: /nix/store/llc6m8bn9vw8y3a4smq70ni6iq752qk4-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-musl/11.3.0/crtbegin.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a PIE object
/nix/store/s22ypd6p0wm9ywhd94fxirvq4bs93d7a-binutils-2.39/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:687: gawk] Error 1
make[2]: Leaving directory '/build/gawk-5.2.1'
make[1]: *** [Makefile:808: all-recursive] Error 1
make[1]: Leaving directory '/build/gawk-5.2.1'
make: *** [Makefile:588: all] Error 2
```

**Comment by trofi** [comment]
> /nix/store/s22ypd6p0wm9ywhd94fxirvq4bs93d7a-binutils-2.39/bin/ld: /nix/store/llc6m8bn9vw8y3a4smq70ni6iq752qk4-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-musl/11.3.0/crtbegin.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a PIE object

That's a PIE/no-PIE mix. if `nixpkgs`'s toolchain default to -fPIE that would explain why `-no-pie` leads to these. Will send a PR in a minute.

**Comment by trofi** [comment]
Proposed the fix as https://github.com/NixOS/nixpkgs/pull/210124

**Pull request merged**